pkg/gui/controllers/local_commits_controller.go

```diff
@@ -504,173 +480,103 @@ func (self *LocalCommitsController) findCommitForQuickStartInteractiveRebase() (
+	// When performing an action that will remove the selected commits, we need to select the
+	// next commit down (which will end up at the start index after the action is performed)
+	if action == todo.Drop || action == todo.Fixup || action == todo.Squash {
+		self.context().SetSelection(startIdx)
+	}
+
+	err := self.c.Git().Rebase.InteractiveRebase(self.c.Model().Commits, startIdx, endIdx, action)
+
 	return self.c.Helpers().MergeAndRebase.CheckMergeOrRebase(err)
 }
 
-// handleMidRebaseCommand sees if the selected commit is in fact a rebasing
+// updateTodos sees if the selected commit is in fact a rebasing
 // commit meaning you are trying to edit the todo file rather than actually
 // begin a rebase. It then updates the todo file with that action
-func (self *LocalCommitsController) handleMidRebaseCommand(action todo.TodoCommand, commit *models.Commit) (bool, error) {
-	if !commit.IsTODO() {
-		return false, nil
-	}
-
-	self.c.LogAction("Update rebase TODO")
-
-	msg := utils.ResolvePlaceholderString(
-		self.c.Tr.Log.HandleMidRebaseCommand,
-		map[string]string{
-			"shortSha": commit.ShortSha(),
-			"action":   action.String(),
-		},
-	)
-	self.c.LogCommand(msg, false)
-
-	if err := self.c.Git().Rebase.EditRebaseTodo(commit, action); err != nil {
-		return false, self.c.Error(err)
+func (self *LocalCommitsController) updateTodos(action todo.TodoCommand, selectedCommits []*models.Commit) error {
+	if err := self.c.Git().Rebase.EditRebaseTodo(selectedCommits, action); err != nil {
+		return self.c.Error(err)
 	}
 
-	return true, self.c.Refresh(types.RefreshOptions{
+	return self.c.Refresh(types.RefreshOptions{
 		Mode: types.SYNC, Scope: []types.RefreshableView{types.REBASE_COMMITS},
 	})
 }
 
-func (self *LocalCommitsController) rebaseCommandEnabled(action todo.TodoCommand) func(*models.Commit) *types.DisabledReason {
-	return func(commit *models.Commit) *types.DisabledReason {
-		if commit.Action == models.ActionConflict {
-			return &types.DisabledReason{Text: self.c.Tr.ChangingThisActionIsNotAllowed}
-		}
-
-		if !commit.IsTODO() {
-			if self.c.Model().WorkingTreeStateAtLastCommitRefresh != enums.REBASE_MODE_NONE {
-				// If we are in a rebase, the only action that is allowed for
-				// non-todo commits is rewording the current head commit
-				if !(action == todo.Reword && self.isHeadCommit()) {
-					return &types.DisabledReason{Text: self.c.Tr.AlreadyRebasing}
-				}
-			}
-
-			return nil
-		}
-
-		// for now we do not support setting 'reword' because it requires an editor
-		// and that means we either unconditionally wait around for the subprocess to ask for
-		// our input or we set a lazygit client as the EDITOR env variable and have it
-		// request us to edit the commit message when prompted.
-		if action == todo.Reword {
-			return &types.DisabledReason{Text: self.c.Tr.RewordNotSupported}
-		}
-
-		if allowed := isChangeOfRebaseTodoAllowed(action); !allowed {
```

Comment:
This line was buggy, but we shouldn't get rid of it (after fixing it). It should have passed `commit.Action` rather than `action`. The point of this check is to make sure that you are not turning a `merge` into a `fixup`, or an `update-ref` into a `squash`, none of which would make sense.

Note, for `merge` it may or may not be useful to turn it into a `drop`, see [this discussion](https://github.com/jesseduffield/lazygit/issues/3164#issuecomment-1858925791). It is currently allowed on master (because of the above bug), fixing this line would disallow it. I'd say it's not a big deal either way, but if we want to allow it, we'd have to pass both the old and the new action to `isChangeOfRebaseTodoAllowed` and special-case this before doing the map lookup.

For `update-ref`, while it's not useful to turn it into anything else, it would be useful to delete it, and the 'd' key would make perfect sense for that (after a confirmation panel, probably). I often wished this were possible. We don't have to worry about it as part of this PR, but it will mean that hitting 'd' on a multi-selection of `update-ref` and other todos should be disallowed even though hitting 'd' on every single item is possible.

Reply:
I've added [`ccac79d` (#3232)](https://github.com/jesseduffield/lazygit/pull/3232/commits/ccac79d46f0534f15798163d1e2b100760b3a07f) and I'm just disallowing any changes to a non-standard TODO action.